rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -901,69 +921,129 @@ void PlayerImpl::load_storage_content()
 {
-  // Note: We need to use message_queue_.peek() instead of message_queue_.try_dequeue(message)
-  // to support play_next() API logic.
-  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = peek_next_message_from_queue();
-  { // Notify play_next() that we are ready for playback
-    // Note: We should do notification that we are ready for playback after peeking pointer to
-    // the next message. message_queue_.peek() is not allowed to be called from more than one
-    // thread concurrently.
+  { // Notify play_next()/seek() that we are ready for playback
     std::lock_guard<std::mutex> lk(ready_to_play_from_queue_mutex_);
     is_ready_to_play_from_queue_ = true;
     ready_to_play_from_queue_cv_.notify_all();
   }
-  while (rclcpp::ok() && !stop_playback_ &&
-    message_ptr != nullptr && !shall_stop_at_timestamp(message_ptr->recv_timestamp))
-  {
-    // Do not move on until sleep_until returns true
-    // It will always sleep, so this is not a tight busy loop on pause
-    while (rclcpp::ok() && !clock_->sleep_until(message_ptr->recv_timestamp)) {
-      if (std::atomic_exchange(&cancel_wait_for_next_message_, false)) {
-        break;
+  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = take_next_message_from_queue();
+
+  // While we haven't stopped playing
+  while (rclcpp::ok() && !stop_playback_) {
+    // While there's a message to play and we haven't reached the end timestamp yet
+    while (rclcpp::ok() && !stop_playback_ &&
+      message_ptr != nullptr && !shall_stop_at_timestamp(message_ptr->recv_timestamp))
+    {
+      // Sleep until the message's replay time, do not move on until sleep_until returns true
+      // It will always sleep, so this is not a tight busy loop on pause
+      // However, skip sleeping if we're trying to play the next message
+      while (rclcpp::ok() && !stop_playback_ && !play_next_.load() &&
+        !clock_->sleep_until(message_ptr->recv_timestamp))
+      {
+        // Stop sleeping if cancelled
+        if (std::atomic_exchange(&cancel_wait_for_next_message_, false)) {
+          break;
+        }
       }
-    }
-    std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
-    if (rclcpp::ok()) {
-      if (skip_message_in_main_play_loop_) {
-        skip_message_in_main_play_loop_ = false;
-        cancel_wait_for_next_message_ = false;
-        message_ptr = peek_next_message_from_queue();
-        continue;
+
+      std::lock_guard<std::mutex> lk_skip_message(skip_message_in_main_play_loop_mutex_);
+      if (rclcpp::ok()) {
+        // This means that the message we took from the queue's top was invalidated after a seek(),
+        // so we need to take a fresh element from the top of the queue, unless we have to stop
+        if (skip_message_in_main_play_loop_) {
+          skip_message_in_main_play_loop_ = false;
+          cancel_wait_for_next_message_ = false;
+          if (stop_playback_.load()) {
+            break;
```

Comment:
If we are exiting by by `stop_playback_ == true` and `play_next_` also `true` if we will not notify `play_next` thread he will stuck and wait forever.
However, it will be better to add such notification at the very end of the function and consider wrapping the main loop in a try-catch to avoid locking the `play_next` thread forever in case of an exception in the main playback loop.
```
  if (play_next_.exchange(false)) {
      std::lock_guard<std::mutex> lk(finished_play_next_mutex_);
      finished_play_next_ = true;
      play_next_result_ = false;
      finished_play_next_cv_.notify_all();
    }
```

Reply:
I would just change it back, because then that situation will get correctly handled

```diff
-          if (stop_playback_.load()) {
-            break;
-          } else {
-            message_ptr = take_next_message_from_queue();
-            continue;
-          }
+         message_ptr = take_next_message_from_queue();
+         continue;
```

Reply:
Okay. We've synched with @christophebedard offline and decided that we don't need wrapping the outer playback loop in try-catch with notification to unblock the play_next thread, and we also don't need to add 
```
    if (play_next_.exchange(false)) {
      std::lock_guard<std::mutex> lk(finished_play_next_mutex_);
      finished_play_next_ = true;
      play_next_result_ = false;
      finished_play_next_cv_.notify_all();
    }
```
at the end of the outer play back loop in this function. Basically, this block is going to be without changes as it is now.
Because we have a try-catch on the upper level in the `play()` method and, we are going to add
```
    if (play_next_.exchange(false)) {
      std::lock_guard<std::mutex> lk(finished_play_next_mutex_);
      finished_play_next_ = true;
      play_next_result_ = false;
      finished_play_next_cv_.notify_all();
    }
```
at the end of the `playback_thread_` right before leaving it.

Also, we will need to add. 
```
  if (!is_in_playback_) {
    RCLCPP_WARN_STREAM(owner_->get_logger(), "Called play next, but not in a playback state.");
    return false;
  }
```
at the beginning of the `bool PlayerImpl::play_next()` method.

Reply:
please see c0fe94dc527d3698c31737c3704cc3c02b19bd64 for the changes.